source/game/field/obj/ObjectRock.cc

```diff
@@ -0,0 +1,169 @@
+#include "ObjectRock.hh"
+
+#include "game/field/CollisionDirector.hh"
+
+#include "game/kart/KartCollide.hh"
+
+namespace Field {
+
+/// @addr {0x8076F2E0}
+ObjectRock::ObjectRock(const System::MapdataGeoObj &params) : ObjectCollidable(params) {}
+
+/// @addr{0x8076F344}
+ObjectRock::~ObjectRock() = default;
+
+/// @addr{0x8076F384}
+void ObjectRock::init() {
+    m_railInterpolator->init(0.0f, 0);
+    m_railInterpolator->setCurrVel(static_cast<f32>(m_mapObj->setting(2)));
+    m_railInterpolator->calc();
+
+    m_state = State::Tangible;
+    m_flags.setBit(eFlags::Position);
+    m_pos.y = m_startYPos = m_railInterpolator->curPos().y;
+
+    EGG::Vector3f curTanDirNorm = m_railInterpolator->curTangentDir();
+    curTanDirNorm.normalise();
+    setMatrixTangentTo(EGG::Vector3f::ey, curTanDirNorm);
+    calcTransform();
+    m_angRad = 0.0f;
+    m_angSpd = INITIAL_ANGULAR_SPEED;
+    m_cooldownTimer = m_mapObj->setting(0);
+    m_colTranslate.x = 0.0f;
+    m_colTranslate.y = static_cast<f32>(m_mapObj->setting(3));
+    m_colTranslate.z = 0.0f;
+    calcTransform();
+}
+
+/// @addr{0x8076F590}
+void ObjectRock::calc() {
+    switch (m_state) {
+    case State::Tangible:
+        calcTangible();
+        break;
+    case State::Intangible:
+        calcIntangible();
+        break;
+    }
+
+    --m_cooldownTimer;
+    EGG::Vector3f scaledTang =
+            m_railInterpolator->curTangentDir() * m_railInterpolator->getCurrVel();
+    m_colTranslate.x = scaledTang.x;
+    m_colTranslate.z = scaledTang.z;
+}
+
+// @addr{0x8076F768}
+void ObjectRock::calcTangible() {
+    auto railStatus = m_railInterpolator->calc();
+    if (railStatus == RailInterpolator::Status::ChangingDirection) {
+        breakRock(EGG::Vector3f::ey);
+    }
+
+    m_flags.setBit(eFlags::Position);
+    m_pos.x = m_railInterpolator->curPos().x;
+    m_flags.setBit(eFlags::Position);
+    m_pos.z = m_railInterpolator->curPos().z;
+    m_colTranslate.y = m_colTranslate.y - 2.0f;
+
+    m_flags.setBit(eFlags::Position);
+    m_pos += m_colTranslate;
+
+    checkSphereFull();
+    calcTangibleSub();
+}
+
+// @addr {0x8076FA60}
+void ObjectRock::checkSphereFull() {
+    CollisionInfo info;
+    info.bbox.setZero();
+
+    EGG::Vector3f offset(0.0f, -(m_scale.x * 240.0f - 50.0f), 0.0f);
+    EGG::Vector3f pos = m_pos + offset;
+
+    if (CollisionDirector::Instance()->checkSphereFull(50.0f, pos, EGG::Vector3f::inf,
+                KCL_TYPE_FLOOR, &info, nullptr, 0)) {
+        m_colTranslate.y *= -0.3f;
+        m_angSpd = std::max((static_cast<f32>(m_mapObj->setting(2)) * 360.0f) /
+                        (m_scale.x * 480.0f * F_PI),
+                m_angSpd + 1.0f);
```

Comment:
asm shows operands in a slightly different order
```suggestion
    if (CollisionDirector::Instance()->checkSphereFull(50.0f, pos, EGG::Vector3f::inf,
                KCL_TYPE_FLOOR, &info, nullptr, 0)) {
        m_colTranslate.y *= -0.3f;
        m_angSpd = std::max(360.0f * static_cast<f32>(m_mapObj->setting(2)) /
                        (480.0f * m_scale.x * F_PI),
                m_angSpd + 1.0f);
```